extension/src/lib/chromeSettings.js

```diff
@@ -0,0 +1,176 @@
+                retryTime: 10
+            }
+        },
+        notifications: {
+            type: "gpii.chrome.notifications"
+        }
+    },
+    model: {
+        settings: "{settings}.options.defaultSettings",  // Defaults
+        promptInstall: true
+    },
+    invokers: {
+        updateSettings: {
+            funcName: "gpii.chrome.settings.updateSettings",
+            args: ["{that}",  "{arguments}.0"]
+        }
+    },
+    listeners: {
+        "{wsConnector}.events.onSettingsChange": "{settings}.updateSettings",
+        "onCreate.bindChromeEvents": "gpii.chrome.settings.bindChromeEvents",
+        "onStartup.updateModel": {
+            changePath: "promptInstall",
+            value: false
+        },
+        "onInstalled.loadExtensionHolders": "{that}.events.onLoadExtensionHolders",
+        "onUpdated.loadExtensionHolders": "{that}.events.onLoadExtensionHolders",
+        "onStartup.loadExtensionHolders": {
+            listener: "{that}.events.onLoadExtensionHolders",
+            priority: "after:updateModel"
+        }
+    },
+    distributeOptions: [{
+        record: {
+            "onExtensionMissing.handle": {
+                funcName: "gpii.chrome.settings.handleExtensionMissing",
+                args: ["{settings}", "{that}"]
+            }
+        },
+        target: "{settings > gpii.chrome.extensionHolder}.options.listeners"
+    }]
+});
+
+gpii.chrome.settings.bindChromeEvents = function (that) {
+    chrome.runtime.onInstalled.addListener(function (object) {
+        if (object.reason === chrome.runtime.OnInstalledReason.INSTALL) {
+            that.events.onInstalled.fire();
+        }
+        if (object.reason === chrome.runtime.OnInstalledReason.UPDATE) {
+            that.events.onUpdated.fire();
+        }
+    });
+
+    chrome.runtime.onStartup.addListener(function () {
+        that.events.onStartup.fire();
+    });
+};
+
+gpii.chrome.settings.updateSettings = function (that, settings) {
+    that.applier.change("settings", settings || that.options.defaultSettings);
+};
+
+gpii.chrome.settings.handleExtensionMissing = function (that, extension) {
+    if (that.model.promptInstall || extension.model.extensionEnabled) {
+        var options = {
+            type: "basic",
+            title: "GPII notifications",
+            message: "To use " + extension.options.settingName + ", please install " + extension.options.name + ".",
+            iconUrl: chrome.extension.getURL("./") + "images/gpii.png",
+            requireInteraction: true,
+            buttons: [{
+                title: "Install from Chrome Web Store"
+            }]
+        };
+
+        that.notifications.create(options, function (id) {
+            var cb = function (notificationId) {
+                if (notificationId === id) {
+                    window.open(extension.options.installationUrl);
+                    that.notifications.events.onButtonClicked.removeListener(cb);
```

Comment:
This workflow is all a bit obscure. Always a smell when we remove listeners in handlers - it seems that there is implicitly an expectation that there is just one notification open at a time, is this really true? It would be nice if we could manage to just install a single handler which routed events to the appropriate target.
Depending on this, it might or might not make sense to create one or more dynamic component instance which form the "backing" for a notification rather than hiding all the state three levels of closure deep like this.

Reply:
Refactored the notification component as a wrapper around a single notification. Using dynamic components to create each instance.